src/http/url.cc

```diff
@@ -68,6 +81,35 @@ bool url_decode(const std::string_view& in, sstring& out) {
     return true;
 }
 
+sstring url_encode(const std::string_view& in) {
+    size_t encodable_chars = 0;
+    for (size_t i = 0; i < in.length(); i++) {
+        if (should_encode(in[i])) {
+            encodable_chars++;
+        }
+    }
+
+    if (encodable_chars == 0) {
+        return sstring(in);
+    }
+
+    sstring ret(in.length() + encodable_chars * 2 + 1, 0);
+    size_t o = 0;
+    for (size_t i = 0; i < in.length(); i++) {
+        if (in[i] == ' ') {
+            ret[o++] = '+';
+        } else if (should_encode(in[i])) {
+            ret[o++] = '%';
+            ret[o++] = char_to_hex(((unsigned char)in[i]) >> 4);
+            ret[o++] = char_to_hex(in[i] & 0xF);
+        } else {
+            ret[o++] = in[i];
+        }
+    }
+    ret.resize(o);
```

Comment:
If you use `%20` instead of `+`, I think you won't need this resize() at the end.